### ISSUE TITLE
Activation functions for complex tensors

### ISSUE THREAD

**Issue by rjkilpatrick** [pytorch/pytorch#47052]
## 🚀 Feature

Activation functions in `torch.nn` should support complex activation functions.

## Motivation

Complex numbers are different from just a simple pair of numbers with their inherent periodicity.
Complex neural networks are a growing research field.
Allowing for these would allow [physical simulation](https://arxiv.org/abs/2009.11810) of complex neural networks with optics.
Complex neural networks need complex activation functions

Similar to https://github.com/pytorch/pytorch/issues/46642.

## Pitch

These activation functions should be implemented for both forward and backward operations on the CPU and CUDA.
The documentation should be clear on whether it calculates the activation function separately for the real and complex parts, or uses the argument and the magnitude.

- [ ] nn.ELU
- [ ] nn.Hardshrink
- [ ] nn.Hardsigmoid
- [ ] nn.Hardtanh
- [ ] nn.Hardswish
- [ ] nn.LeakyReLU
- [ ] nn.LogSigmoid
- [ ] nn.MultiheadAttention
- [ ] nn.PReLU
- [ ] nn.ReLU
- [ ] nn.ReLU6
- [ ] nn.RReLU
- [ ] nn.SELU
- [ ] nn.CELU
- [ ] nn.GELU
- [ ] nn.Sigmoid
- [ ] nn.SiLU
- [ ] nn.Softplus
- [ ] nn.Softshrink
- [ ] nn.Softsign
- [ ] nn.Tanh
- [ ] nn.Tanhshrink
- [ ] nn.Threshold

## Additional context

See <https://giggleliu.github.io/2018/02/01/complex_bp.html> for an explanation on how complex numbers are deeper than mere pairs of real numbers.

**Comment by albanD** [comment]
Hi,

Thanks for the feature request!
For reference, we have https://github.com/pytorch/pytorch/issues/46374 that tracks support for complex numbers across `torch.nn`.

The main question I would have is about the choice between "the real and complex parts, or uses the argument and the magnitude.".
As replacing what a Module computes in torch.nn is tricky, we want to make sure that we make the right choice here.

In particular, would you have references or examples of code that uses such activation that could motivate the choice?

**Comment by rjkilpatrick** [comment]
When you are extending a real activation function `f` (e.g. ReLU) to complex numbers `z = a + ib = re^(iθ)`, in general there are two ways:
1. `f(z) = f(a) + i f(b)`, as a real-imaginary activation function, i.e.:
``` python
import torch
from torch.nn.functional import relu

def real_imaginary_relu(z):
    return relu(z.real) + 1.j * relu(z.imag)
```
Where the two components are handled separately. This activation function is consistent with the currently implemented functions, as real numbers are just `a + 0i`. However, when the sign is different between the two parts of the complex number, it doesn't satisfy the [Cauchy–Riemann equations](https://en.wikipedia.org/wiki/Cauchy%E2%80%93Riemann_equations).

2. As a phase-amplitude activation e.g. `f(z) = f(|r|) * e^iθ` (from [Complex-Valued Neural Networks](http://link.springer.com/10.1007/978-3-642-27632-3)):
``` python
def phase_amplitude_relu(z):
    return relu(torch.abs(z)) * torch.exp(1.j * torch.angle(z))
```
Where the phase and the amplitude of the complex number are handled separately (i.e. the real and imaginary components are coupled in some way). This is useful for considering [waves](https://farside.ph.utexas.edu/teaching/315/Waves/node88.html) but can be confusing because -9 is considered "greater" than 2 (as -9 = 9 * exp(iπ)), see [math stackexhange](https://math.stackexchange.com/questions/1116022/can-a-complex-number-ever-be-considered-bigger-or-smaller-than-a-real-number). This way the phase propogates through the network.

The [motivation](https://arxiv.org/abs/1503.03438) behind complex numbers neural networks is effectively the periodicity introduced. I.e. multiplication by 'i' is equivalent to rotation by 90 degrees, which is half the transformation to '-1'. To ensure this periodicity, I would prefer (2.) (see Deep Complex Networks below)
As an aside, this would also mean it may be more efficient to store the numbers in polar form.

There are more sophisticated extensions of real activation functions, with their reasoning and extent detailed in 3.4 of **[Deep Complex Networks](https://arxiv.org/abs/1705.09792)**.

e.g. modReLU
``` python
class mod_relu(torch.nn.Module):
    def __init__(self, in_features):
        super().__init__()
        self.b = Parameter(torch.tensor(0.))
        self.b.requiresGrad = True

    def forward(self, x):
        return relu(torch.abs(z) + b) * torch.exp(1.j * torch.angle(z)) 
```

**Comment by mpariente** [comment]
From what I understood, the Cauchy-Riemann equations doesn't need to be satisfied (https://pytorch.org/docs/master/notes/autograd.html#what-are-complex-derivatives).

Intuitively, I'd choose (2) over (1). But (1) seems more non-linear than (2), and maybe it would be a good thing. 
IMO, it's too early to choose because the community won't agree. 

Would it be possible/feasible to provide wrappers for activations for both (1) and (2)? We have one for (1) [here](https://github.com/mpariente/asteroid/blob/master/asteroid/complex_nn.py#L69) for example.

**Comment by albanD** [comment]
I guess you can write a generic wrapper like this?
```python
class ComplexAct(nn.Module):
    def __init__(self, act, use_phase=False):
        # act can be either a function from nn.functional or a nn.Module if the
        # activation has learnable parameters
        self.act = act
        sefl.use_phase = use_phase

    def forward(self, inp):
        if self.use_phase:
            return self.act(torch.abs(z)) * torch.exp(1.j * torch.angle(z)) 
        else:
            return self.act(z.real) + 1.j * self.act(z.imag)
```

**Comment by mpariente** [comment]
Yes, something like that is completely possible, and simple to implement in downstream projects. 
What do you think, should it be implemented in downstream projects, or in PyTorch?

**Comment by ezyang** [comment]
For me, it comes down to a matter of how "established" things are. A bad outcome would be for PyTorch to pick one of two mutually exclusive choices to put into its library, and then the community going down the other road, leaving PyTorch with an outdated API that it can't update (because BC) which isn't actually useful for anyone in practice. That being said, when things are easy or obvious, there isn't any reason we shouldn't put it in PyTorch.

**Comment by albanD** [comment]
@ezyang what do you think about adding a cleaned up version of the `ComplexAct` above in `torch.nn.utils`. It will allow people to experiment easily without having to rely on third party libs and does not block us in any choice.

**Comment by ezyang** [comment]
IDK. It's so short that people can just copy paste it into their projects if they want...

**Comment by anjali411** [comment]
@rjkilpatrick thanks for the feature request. It seems that there is no community consensus on how the activation functions should look like for complex numbers. We are constantly looking out for research developments in this field to add support for complex numbers but I think we are wary of adding anything in `PyTorch` right now since things are still up in the air.

**Comment by e13000** [comment]
I work in audio signal processing and the community mostly uses (1). In (2), I dont think relu has any effect since torch.abs(z) is already non-negative (am I missing anything here?).

**Comment by rjkilpatrick** [comment]
I'll implement an extension with the likely candidates, with a PyTorch compatible license. If any of them become useful, I'll submit a PR.
This way PyTorch doesn't commit to any standard for complex losses and hopefully gets code quickly when a standard emerges.

**Comment by mruberry** [comment]
I'm exciting to see what the community develops on top of the support that PyTorch has for complex modules, @rjkilpatrick.

**Comment by soumickmj** [comment]
I started working on a package for Complex Module for PyTorch. The usage will be similar to PyTorch. Will just need to replace torch with torchcomplex.
Please have a look if you find it usefull.
Comments, feedbacks are most appreciated. 
Also, I would welcome you all for collaborations. 
You can find it here: https://github.com/soumickmj/pytorch-complex

**Comment by tygert** [comment]
Might you want to consider including a standardized implementation of the absolute value of complex numbers, calling it "nn.Abs" or something like that? That may be the most commonly used activation in complex-valued neural nets, as in Stephane Mallat's and Joan Bruna's scattering networks, or the work of Joan Bruna, Soumith Chintala, Yann LeCun, et al. available at https://arxiv.org/abs/1503.03438 ... a polar decomposition into both the absolute value and the complex argument (angle) could be another possibility.

**Comment by soumickmj** [comment]
@tygert Thanks. I did not add abs because of torch.abs. But I will add one, that would be fairly straightforward. 
Regarding the polar decomposition, you wanted this as an activation function?
If yes, how that will be working? From the complex input, we can decompose as abs and angle. But, then? Will this activation be giving two separate output? I'm not sure exactly about this.

**Comment by tygert** [comment]
@soumickmj : nn.Abs or something like that would be my personal ideal. I could see how someone might want a decomposition into the absolute value and the angle; maybe that someone could handle that if the time comes?

**Comment by arashbehi** [comment]
To resurrect this topic, here in [ the second implementation to preserve phase propagation](https://github.com/pytorch/pytorch/issues/47052#issuecomment-718948801), the `backward` function needs to be implemented right?

**Comment by anjali411** [comment]
> the second implementation to preserve phase propagation

yeah

**Comment by egrinstein** [comment]
Hi all,

I just wanted to point out that

> ```python
> def phase_amplitude_relu(z):
>     return relu(torch.abs(z)) * torch.exp(1.j * torch.angle(z))
> ```

is in fact a linear activation function, since applying a ReLU
to the amplitude of a complex number, which is always non-negative, does not alter its value.

An alternative provided in [Complex-Valued Neural Networks](http://link.springer.com/10.1007/978-3-642-27632-3)
is substituting the `relu` for the `tanh` function.

**Comment by SantaTitular** [comment]
Hi all,

Apologies for ressurecting this topic as well, but I think that currently, it is possible to build complex-valued neural networks using pytorch modules. Where can I find documentation regarding activation functions for complex-valued data, e.g., reLu, tanh, etc.? Is it more appropriate to use lightweight extensions as suggested in #46374 ?
